### PR TITLE
Do not print emojis when --no-color is set

### LINKED ISSUE #301: --no-color flag or envar also removes emojis
### Description of problem

We have a mechanism to remove text decoration (color or styling such as bolding) to support machine readers such as assistive technology or automation, or user preferences.  

The recently added emojis are not suppressed by passing the `--no-color` flag or setting the `NO_COLOR` envar.

### Proposed solution

```gherkin
Given I pass a --no-color flag, or set the NO_COLOR envar
When I run a command
Then I should see plain / undecorated text
```

#### Help Text 
Current

```
      --no-color          disable color output in terminals
```

Proposed

```
      --no-color          deactivate color, bold, and emoji output
```

### Example

<Code snippets that illustrate the when/then blocks>

### Describe alternatives you've considered
An emoji-specific flag `--no-emoji` 

Supporting multiple toggles
```
--no-color  // remove color and styling only, leave emojis
--no-emoji  // remove emojis only, leave color and styling
--no-ansi   // remove all decoration 
```

### PR DESCRIPTION
**Pull request by warango4**
Pull request

### What this PR does / why we need it
Avoid printing emojis when either flag `--no-color` or env var `NO_COLOR` is set.

### Which issue(s) this PR fixes

Fixes #301

### Describe testing done for PR

- Updated unit tests
- Updated integration tests

### CODE REVIEW

**Comment by warango4** [comment]
image

**Comment by warango4** [comment]
image

File: pkg/commands/workload_get.go [hunk 1]
```diff
@@ -114,6 +115,10 @@ func (opts *WorkloadGetOptions) Exec(ctx context.Context, c *cli.Config) error {
 		return nil
 	}
 
+	if cli.CommandFromContext(ctx).Flags().Changed(cli.StripDash(flags.NoColorFlagName)) {
```
Comment (shaheerkootteeri) [inline pkg/commands/workload_get.go:118]: Can we set these NoColor flag to config in main ?

Comment (warango4) [reply pkg/commands/workload_get.go]: Thanks for your suggestion, I already updated the PR

**Comment by warango4** [comment]
image

File: testing/e2e/workload_test.go [hunk 2]
```diff
@@ -304,18 +304,18 @@ func TestCreateFromGitWithAnnotations(t *testing.T) {
 			},
 		},
 		{
-			Name:         "Get the updated workload",
+			Name:         "Get the updated workload with no color",
```
Comment (shashwathi) [inline testing/e2e/workload_test.go:307]: remove e2e test

**Review by shashwathi — CHANGES_REQUESTED** [review]

File: cmd/plugin/apps/main.go [hunk 3]
```diff
@@ -101,9 +101,12 @@ func main() {
 	}
 
 	cobra.OnInitialize(func() {
-		// sync survey and faith option to disable color
+		// sync survey and fatih option to disable color
```
[the remark this answers is not in the document]
Comment (warango4) [reply cmd/plugin/apps/main.go:104]: github.com/fatih/color

File: pkg/cli-runtime/config.go [hunk 4]
```diff
@@ -112,10 +108,22 @@ func (c *Config) Boldf(format string, a ...interface{}) (n int, err error) {
...
+	}
 	emojiFormat := fmt.Sprintf("%s%s%s", string(emoji), " ", format)
 	return printer.BoldColor.Fprintf(c.Stdout, emojiFormat, a...)
 }
 
+func (c *Config) Emoji(emoji Icon, format string) string {
+	if c.NoColor {
+		return format
+	}
+	emojiFormat := fmt.Sprintf("%s%s%s", string(emoji), " ", format)
```
Comment (odinnordico) [inline pkg/cli-runtime/config.go:123]: use the blank space in the `Sprintf` format instead of as parameter

**Review by odinnordico — CHANGES_REQUESTED** [review]

File: pkg/cli-runtime/config_test.go [hunk 5]
```diff
@@ -160,26 +160,29 @@ func TestConfig_EmojiPrint(t *testing.T) {
...
 	}{{
-		name:    "EmojiSuccessf",
-		icon:    cli.ThumpsUp,
-		args:    []interface{}{"Pods created Successfully"},
-		printer: config.EmojiSuccessf,
-		stdout:  `👍 Pods created Successfully`,
-	}, {
 		name:    "EmojiBoldf",
 		icon:    cli.FloppyDisk,
+		noColor: false,
```
Comment (shashwathi) [inline pkg/cli-runtime/config_test.go:170]: by default bool takes false value. Why is this set explicitly?

File: pkg/commands/workload_get.go [hunk 6]
```diff
@@ -115,14 +115,16 @@ func (opts *WorkloadGetOptions) Exec(ctx context.Context, c *cli.Config) error {
 	}
 
 	//print workload details
-	c.EmojiBoldf(cli.Antenna, "Overview\n")
+	c.Boldf(c.Emoji(cli.Antenna, "Overview\n"))
+	//c.EmojiBoldf(cli.Antenna, "Overview\n")
```
Comment (shashwathi) [inline pkg/commands/workload_get.go:119]: remove commented line

Comment (shashwathi) [reply pkg/commands/workload_get.go:119]: for all the prints in this file

File: testing/e2e/workload_test.go [hunk 7]
```diff
@@ -392,23 +392,6 @@ func TestCreateFromGitWithAnnotations(t *testing.T) {
...
-			Name:         "Get the updated workload",
-			WorkloadName: "test-create-git-annotations-workload",
-			Command: *it.NewTanzuAppsCommandLine(
-				"workload", "get", "test-create-git-annotations-workload", namespaceFlag),
-			Verify: func(t *testing.T, output string, err error) {
-				if regexPod.FindString(output) == "" {
-					t.Errorf("expected Pod results in output %v", output)
-					t.FailNow()
-				}
-				decodedString := strconv.QuoteToASCII(output)
```
Comment (shashwathi) [inline testing/e2e/workload_test.go:405]: when you removed this test, remove `strconv` as dep as well. tests are failing because of that

**Review by shashwathi — COMMENTED** [review]
- need e2e tests
- Cleanup commented code workload get file.

File: pkg/cli-runtime/config_test.go [hunk 8]
```diff
@@ -153,40 +153,34 @@ func TestConfig_Print(t *testing.T) {
...
+		input:  "Source",
+		output: `💾 Source`,
 	}, {
-		name:    "EmojiBoldf",
-		icon:    cli.FloppyDisk,
-		args:    []interface{}{"Source"},
-		printer: config.EmojiBoldf,
-		stdout:  `💾 Source`,
+		name:    "Text without emoji",
+		noColor: true,
```
Comment (shaheerkootteeri) [inline pkg/cli-runtime/config_test.go:174]: We would need an icon too part of the test to make sure the noColor flag working

File: pkg/commands/workload_get.go [hunk 9]
```diff
@@ -197,7 +197,7 @@ func (opts *WorkloadGetOptions) Exec(ctx context.Context, c *cli.Config) error {
 
 	// Print workload issues
 	c.Printf("\n")
-	c.EmojiBoldf(cli.SpeechBalloon, "Messages\n")
+	c.Boldf(c.Emoji(cli.SpeechBalloon, "Messages\n"))
```
Comment (shaheerkootteeri) [inline pkg/commands/workload_get.go:200]: Can this be reversed to avoid any abnormal output on Boldf function because of the rune

**Review by shaheerkootteeri — CHANGES_REQUESTED** [review]

File: docs/command-reference/tanzu_apps_workload_apply.md [hunk 10]
```diff
@@ -73,12 +73,11 @@ tanzu apps workload apply --file workload.yaml
...
       --kubeconfig file   kubeconfig file (default is $HOME/.kube/config)
-      --no-color          disable color output in terminals
+      --no-color          deactivate color, bold, and emoji output
   -v, --verbose int32     number for the log level verbosity (default 1)
 ```
 
 ### SEE ALSO
 
 * [tanzu apps workload](tanzu_apps_workload.md)	 - Workload lifecycle management
-* [environment variable](../working-with-workloads.md#env-vars) support
```
Comment (odinnordico) [inline docs/command-reference/tanzu_apps_workload_apply.md:83]: Why is this removed as part of this PR?

Comment (warango4) [reply docs/command-reference/tanzu_apps_workload_apply.md:83]: Maybe it happened when I ran `make docs` or when rebasing from `main`, I'll remove it. Thanks for pointing this out.

File: testing/e2e/workload_test.go [hunk 11]
```diff
@@ -54,7 +57,8 @@ var (
 func TestCreateFromGitWithAnnotations(t *testing.T) {
 	testSuite := it.CommandLineIntegrationTestSuite{
 		{
-			Name:         "Create workload with valid name from a git repo and annotations",
+			Name: "Create workload with valid name from a git repo and annotations",
+			//Focus:        true,
```
Comment (odinnordico) [inline testing/e2e/workload_test.go:61]: Remove `Focus` comment

File: testing/e2e/workload_test.go [hunk 12]
```diff
@@ -406,11 +410,27 @@ func TestCreateFromGitWithAnnotations(t *testing.T) {
...
+			Name:         "Get the updated workload with color",
+			Skip:         runtime.GOOS == "windows",
 			WorkloadName: "test-create-git-annotations-workload",
 			Command: *it.NewTanzuAppsCommandLine(
 				"workload", "get", "test-create-git-annotations-workload", namespaceFlag),
-			Verify: func(t *testing.T, output string, err error) {
+			Prepare: func(ctx context.Context, t *testing.T) (context.Context, error) {
+				r, w, err := pty.Open()
+				if err != nil {
+					t.Errorf("error while creating pipe %v", err)
```
Comment (odinnordico) [inline testing/e2e/workload_test.go:421]: Update the error message as it is not creating pipes

File: testing/e2e/workload_test.go [hunk 13]
```diff
@@ -421,6 +441,82 @@ func TestCreateFromGitWithAnnotations(t *testing.T) {
 					t.FailNow()
 				}
 			},
+			CleanUp: func(ctx context.Context, t *testing.T) error {
+				ctx = context.WithValue(ctx, "reader", nil)
```
Comment (odinnordico) [inline testing/e2e/workload_test.go:445]: is it necessary to set values to `nil` as `ctx` is not return?

File: testing/suite/command_line.go [hunk 14]
```diff
@@ -107,3 +109,50 @@ func (c *CommandLine) Exec() error {
 
 	return nil
 }
+
+func (c *CommandLine) ExecWithCustomPipe(w *os.File, r *os.File) error {
```
Comment (odinnordico) [inline testing/suite/command_line.go:113]:
```suggestion
func (c *CommandLine) ExecWithCustomPipe(w, r *os.File) error {
```

File: docs/command-reference/tanzu_apps_workload_apply.md (hunk 10 above)
Comment (warango4) [reply docs/command-reference/tanzu_apps_workload_apply.md:83]: This is a problem with `make docs` command. Every time I run it, this line is getting erased. And I have to it because of the change to `no-color` explanation.

File: testing/e2e/workload_test.go (hunk 13 above)
Comment (warango4) [reply testing/e2e/workload_test.go:445]: You're right, I thought I needed to clean up the context every time but it seems it's not necessary. Thanks
